src/Crawler/RepositoryCrawler.cs

```diff
@@ -0,0 +1,208 @@
+﻿using ModelRepoBrowser.Crawler.XmlModels;
+using ModelRepoBrowser.Models;
+using System.Collections.Concurrent;
+
+namespace ModelRepoBrowser.Crawler
+{
+    public class RepositoryCrawler
+    {
+        private static readonly HttpClient client = new();
+        private IDictionary<string, Repository> modelRepositories = new ConcurrentDictionary<string, Repository>();
+
+        /// <summary>
+        /// Parse the Repository tree from the root repository following subsidiary links.
+        /// </summary>
+        /// <param name="rootRepositoryUri">The <see cref="Uri"/> of the repository at the root of the model repository tree.</param>
+        /// <returns></returns>
+        public async Task<IDictionary<string, Repository>> CrawlModelRepositories(Uri rootRepositoryUri)
+        {
+            await CrawlRepositories(rootRepositoryUri, null).ConfigureAwait(false);
+            return modelRepositories;
+        }
+
+        private async Task<Repository?> CrawlRepositories(Uri rootRepositoryUri, Repository? parent)
+        {
+            rootRepositoryUri = await PreferHttpsIfAvailable(rootRepositoryUri).ConfigureAwait(false);
+
+            if (modelRepositories.TryGetValue(rootRepositoryUri.Host, out var existingRepository))
+            {
+                if (parent != null)
+                {
+                    existingRepository.ParentSites.Add(parent);
+                }
+
+                return existingRepository;
+            }
+
+            var (root, subsidiarys) = await AnalyseRepository(rootRepositoryUri, parent).ConfigureAwait(false);
+            if (root is not null && modelRepositories.TryAdd(rootRepositoryUri.Host, root))
+            {
+                var subsidiarySites = new HashSet<Repository>();
+                foreach (var subsidiary in subsidiarys)
+                {
+                    var subsidiaryRepository = await CrawlRepositories(subsidiary, root).ConfigureAwait(false);
+                    if (subsidiaryRepository != null)
+                    {
+                        subsidiarySites.Add(subsidiaryRepository);
+                    }
+                }
+
+                root.SubsidiarySites = subsidiarySites;
+            }
+
+            return root;
+        }
+
+        internal async Task<(Repository? Repository, IEnumerable<Uri> SubsidiaryRepositories)> AnalyseRepository(Uri repositoryUri, Repository? parent = null)
+        {
+            try
+            {
+                var ilisiteTask = ParseIlisite(repositoryUri);
+                var modelsTask = CrawlIlimodels(repositoryUri);
+                var metadatasTask = CrawlIlidata(repositoryUri);
+
+                var ilisite = await ilisiteTask.ConfigureAwait(false);
+                var models = await modelsTask.ConfigureAwait(false);
+                var metadatas = await metadatasTask.ConfigureAwait(false);
```

Comment:
REQ: Die Definition der einzelnen Tasks weglassen und gleich die Methoden `await`en. Beispiel: `var ilisite = await ParseIlisite(repositoryUri).ConfigureAwait(false);`

Reply:
@danjov Hier war die Idee, dass die 3 Files pro Repository parallel gehohlt & geparsed werden können. Weshalb muss das hier anders gelöst werden?

Reply:
@Philippluca Ah, alles klar. Ich dachte wir müssen das in einer spezifischen Reihenfolge machen (z.B. ilidata erst nachdem alle _models_ gelesen wurden).